File: src/config/config.test.ts
import getConfiguration from './config';

describe('config loading behavior', () => {
  it('should remove trailing slashes', () => {
    process.env.ENCORE_URL = 'http://encore-instance.io/';
    process.env.CALLBACK_LISTENER_URL = 'http://callback.com';
    process.env.ASSET_SERVER_URL = 'http://assethost.io/';
    process.env.AD_SERVER_URL = 'http://adserver.com';
    process.env.REDIS_URL = 'http://redis.com';
    process.env.REDIS_CLUSTER = 'true';
    process.env.OUTPUT_BUCKET_URL = 's3://ads/';
    process.env.OSC_ACCESS_TOKEN = 'token';
    process.env.KEY_FIELD = 'Url';
    process.env.KEY_REGEX = '[a-zA-Z]';
    process.env.ENCORE_PROFILE = 'test-profile';
    process.env.JIT_PACKAGING = 'true';
    process.env.PACKAGING_QUEUE = 'ad-packaging';
    process.env.ROOT_URL = 'http://eyevinn.ad-normalizer.osaas.io';

    const config = getConfiguration();
    const expectedBucketName = 'ads';
    const expectedEncoreUrl = 'http://encore-instance.io';
    // Assert
    expect(config.bucket).toEqual(expectedBucketName);
    expect(config.encoreUrl).toEqual(expectedEncoreUrl);
    expect(config.keyRegex).toEqual('[a-zA-Z]');
    expect(config.keyField).toEqual('url');
    expect(config.rediscluster).toEqual(true);
    expect(config.encoreProfile).toEqual('test-profile');
    expect(config.jitPackaging).toEqual(true);
    expect(config.packagingQueueName).toEqual('ad-packaging');
    expect(config.rootUrl).toEqual('http://eyevinn.ad-normalizer.osaas.io');
  });
});

File: internal/config/config_test.go
package config

import (
	"testing"

	"github.com/matryer/is"
)

func TestReadConfig(t *testing.T) {
	is := is.New(t)
	configVars := []struct {
		name  string
		value string
	}{
		{"ENCORE_URL", "http://demo-encore.osaas.io"},
		{"REDIS_URL", "redis://demo-valkey.osaas.io"},
		{"REDIS_CLUSTER", "true"},
		{"AD_SERVER_URL", "http://test-ad-server.osaas.io"},
		{"OUTPUT_BUCKET_URL", "s3://test-bucket.osaas.io"},
		{"KEY_FIELD", "url"},
		{"KEY_REGEX", "^[^a-zA-Z0-9]"},
		{"ENCORE_PROFILE", "ad-profile"},
		{"ASSET_SERVER_URL", "http://test-asset-server.osaas.io"},
		{"JIT_PACKAGE", "true"},
		{"ROOT_URL", "http://ad-normalizer.osaas.io"},
		{"PACKAGING_QUEUE", "normalizer-package"},
		{"IN_FLIGHT_TTL", "10"},
	}
	for _, v := range configVars {
		t.Setenv(v.name, v.value)
	}
	config, err := ReadConfig()
	is.NoErr(err)
	is.Equal(config.EncoreUrl.String(), "http://demo-encore.osaas.io")
	is.Equal(config.ValkeyUrl, "redis://demo-valkey.osaas.io")
	is.Equal(config.AdServerUrl.String(), "http://test-ad-server.osaas.io")
	is.Equal(config.BucketUrl.String(), "s3://test-bucket.osaas.io")
	is.Equal(config.InFlightTtl, 10)
	is.Equal(config.KeyField, "url")
	is.Equal(config.KeyRegex, "^[^a-zA-Z0-9]")
	is.Equal(config.EncoreProfile, "ad-profile")
	is.Equal(config.ValkeyCluster, true)
}
